### PR TITLE
Add concurrency control functionality

### PR DESCRIPTION
**Pull request by wangch079**
## Closes https://github.com/elastic/enterprise-search-team/issues/4157

This PR:
1. adds new instance variable `_seq_no` and `_primary_field` to `ESDocument` class
2. makes the `reload` method of `ESDocument` class update the instance variables directly
3. removes `@setter` methods, and makes `prepare` and `validate_filtering` methods of `Connector` class to `reload`.
4. makes the `get_all_docs` method of `ESIndex` to retrieve `_seq_no` and `_primary_field`
5. makes the `update` method of `ESIndex` to accept `if_seq_no` and `if_primary_field` paremeters

#### Pre-Review Checklist
- [x] Covered the changes with automated tests
- [x] Tested the changes locally
- [x] Added a label for each target release version (example: `v7.13.2`, `v7.14.0`, `v8.0.0`)
- [ ] Considered corresponding documentation changes
- [ ] Contributed any configuration settings changes to the configuration reference

### CODE REVIEW

File: connectors/sync_job_runner.py [hunk 1]
```diff
@@ -144,7 +144,11 @@ async def _sync_done(self, sync_status, result=None, sync_error=None):
         else:
             await self.sync_job.done(ingestion_stats=ingestion_stats)
 
-        self.sync_job = await self.sync_job.reload()
+        try:
+            await self.sync_job.reload()
+        except Exception as e:
+            logger.error(f"Failed to reload sync job {self.job_id}. Error: {e}")
+            self.sync_job = None
```
Comment (tarekziade) [inline connectors/sync_job_runner.py:151]: putting `None` here seems pretty drastic, what happens next on the sync_doc call?

**Review by tarekziade — COMMENTED** [review]
just one question about what happens when `reload` fails

File: connectors/sync_job_runner.py (hunk 1 above)
Comment (wangch079) [reply connectors/sync_job_runner.py:151]: ~For now, this is the only place `reload` is used.~

You just reminds me I need to add `reload` in `prepare` and `validate_filtering` methods: https://github.com/elastic/connectors-python/pull/583/commits/e269d684601249972e37979c2d74cf9ce2259f66

>what happens next on the sync_doc call?

It will updates `last_sync_status` to `error`:
https://github.com/elastic/connectors-python/blob/e269d684601249972e37979c2d74cf9ce2259f66/connectors/byoc.py#L483

This is one of the edge cases we need to handle in https://github.com/elastic/connectors-python/pull/517 during job execution.

Comment (wangch079) [reply connectors/sync_job_runner.py:151]: ~For now, this is the only place `reload` is used.~

You just reminds me I need to add `reload` in `prepare` and `validate_filtering` methods: https://github.com/elastic/connectors-python/pull/583/commits/e269d684601249972e37979c2d74cf9ce2259f66

>what happens next on the sync_doc call?

It will updates `last_sync_status` to `error`:
https://github.com/elastic/connectors-python/blob/e269d684601249972e37979c2d74cf9ce2259f66/connectors/byoc.py#L483

This is one of the edge cases we need to handle in https://github.com/elastic/connectors-python/pull/517 during job execution.

**Comment by artem-shelkovnikov** [comment]
@wangch079 I'm looking at the PR and understand individual pieces of it, but don't understand the high-level approach taken. Is there a document/code snippets/diagrams that I could check to understand how it will work when multiple parties try to update the same record?

**Comment by wangch079** [comment]
>@wangch079 I'm looking at the PR and understand individual pieces of it, but don't understand the high-level approach taken. Is there a document/code snippets/diagrams that I could check to understand how it will work when multiple parties try to update the same record?

Hi @artem-shelkovnikov , the implementation doc is [here](https://docs.google.com/document/d/1D3CeQQcDZkHsuGYOpfoVoXVFphlW27FpT_Rifw52wD0/edit#). 

This PR is just to add the concurrency control to the `update` method (make it accept optional params `if_seq_no` and `if_primary_term`), but the real concurrency control is not used until https://github.com/elastic/enterprise-search-team/issues/4159

File: connectors/byoc.py [hunk 2]
```diff
@@ -389,23 +389,10 @@ class Connector(ESDocument):
...
 
-    @status.setter
-    def status(self, value):
-        if isinstance(value, str):
-            value = Status(value)
-        if not isinstance(value, Status):
-            raise TypeError(value)
-
-        self._source["status"] = value
-
```
Comment (artem-shelkovnikov) [inline connectors/byoc.py:400]: So to double check, setters were removed because they are not used?

File: connectors/byoc.py [hunk 3]
```diff
@@ -575,6 +552,7 @@ async def prepare(self, config):
             raise ConnectorUpdateError(
                 f"Could not update service type/configuration for connector {self.id}"
             )
+        await self.reload()
```
Comment (artem-shelkovnikov) [inline connectors/byoc.py:555]: What's the reason of adding the reload here?

File: connectors/byoc.py [hunk 4]
```diff
@@ -601,8 +579,8 @@ async def validate_filtering(self, validator):
                 if validation_result.state == FilteringValidationState.VALID:
                     filter_["active"] = filter_.get("draft")
 
-        self.filtering = filtering
         await self.index.update(doc_id=self.id, doc={"filtering": filtering})
+        await self.reload()
```
Comment (artem-shelkovnikov) [inline connectors/byoc.py:583]: Same - do we actually want to reload everywhere then?

File: connectors/es/index.py [hunk 5]
```diff
@@ -49,17 +57,25 @@ async def fetch_by_id(self, doc_id):
...
-        return self._create_object(resp.body)
+        return resp.body
 
     async def index(self, doc):
         resp = await self.client.index(index=self.index_name, document=doc)
         return resp["_id"]
 
-    async def update(self, doc_id, doc):
-        await self.client.update(index=self.index_name, id=doc_id, doc=doc)
+    async def update(self, doc_id, doc, if_seq_no=None, if_primary_term=None):
```
Comment (artem-shelkovnikov) [inline connectors/es/index.py:71]: So for now `if_seq_no` is not used, should it be added in a different PR, or should logic that uses it be added here?

**Review by artem-shelkovnikov — COMMENTED** [review]
I look at the changeset and don't quite understand - nothing seem to functionally change regarding concurrency control in this PR. I see `if_seq_no` added, but not used - should this change be broken into multiple ones instead? 
So changes about `DocumentNotFoundError` can be a different PR. Changes about `reload` - are they relevant to `if_seq_no` change?

I'm pretty confused tbh.

File: connectors/byoc.py (hunk 3 above)
Comment (wangch079) [reply connectors/byoc.py:555]: So the `connector` object has the latest `service_type` and `configuration`.

File: connectors/byoc.py (hunk 4 above)
Comment (wangch079) [reply connectors/byoc.py:583]: Not everywhere. We try not to do so if possible. But this place is critical. Imagine a customer edits filtering in Kibana, BE validates it, and updates the active filtering. If we don't reload here, the connector will still hold the out-dated filtering, which will be used to schedule a new sync job.

File: connectors/es/index.py (hunk 5 above)
Comment (wangch079) [reply connectors/es/index.py:71]: I'm trying to break the changes as small as possible. The purpose of this PR is, as the title says, to `Add concurrency control functionality`, i.e. make it capable of using optimistic concurrency control, but not necessarily use it.

**Comment by wangch079** [comment]
Close this PR as it has been further broken down into
1. #607
2. #608 
3. #609 
4. #610 
5. #611
